extensions/ql-vscode/src/common/vscode/file-path-discovery.ts

```diff
@@ -0,0 +1,190 @@
+import { Discovery } from "../discovery";
+import {
+  EventEmitter,
+  RelativePattern,
+  Uri,
+  WorkspaceFoldersChangeEvent,
+  workspace,
+} from "vscode";
+import { MultiFileSystemWatcher } from "./multi-file-system-watcher";
+import { AppEventEmitter } from "../events";
+import { extLogger } from "..";
+import { FilePathSet } from "../file-path-set";
+import { exists, lstat } from "fs-extra";
+import { containsPath } from "../../pure/files";
+import { getOnDiskWorkspaceFoldersObjects } from "./workspace-folders";
+
+interface PathData {
+  path: string;
+}
+
+/**
+ * Discovers all files matching a given filter contained in the workspace.
+ *
+ * Scans the whole workspace on startup, and then watches for changes to files
+ * to do the minimum work to keep up with changes.
+ *
+ * Can configure which changes it watches for, which files are considered
+ * relevant, and what extra data to compute for each file.
+ */
+export abstract class FilePathDiscovery<T extends PathData> extends Discovery {
+  /** The set of known paths we are tracking */
+  protected paths: T[] = [];
+  protected readonly onDidChangePathsEmitter: AppEventEmitter<void>;
+
+  private readonly changedFilePaths = new FilePathSet();
```

Comment:
I would find it helpful to have a comment here explaining how this variable works since you now need to know how to find that information.

Reply:
Good point. I've added more documentation to all of the fields in this class.

For this specific field, what I've written is:
```
  /**
   * The set of file paths that may have changed on disk since the last time
   * refresh was run. Whenever a watcher reports some change to a file we add
   * it to this set, and then during the next refresh we will process all
   * file paths from this set and update our internal state to match whatever
   * we find on disk (i.e. the file exists, doesn't exist, computed data has
   * changed).
   */
```

Reply:
Now that I've thought about this now, I'm thinking of removing the custom `FilePathSet` class and instead using a regular `Set`.

The reasons for using a set rather than an array is to remove duplicates. If a file path has changed twice since we last ran `refresh` then we don't need to process that file twice. We can just process it once and deal with whatever state we find on disk.

The reason I went with this `FilePathSet` that avoids having child paths in the set, is that for some of the file watcher events it only reports a directory. I think only the case where this happens is when a directory is added or deleted. So if a directory and a subdirectory are deleted then we'd get "duplicate" events because we could get the same effect by just processing the parent directory.

I'm coming around to thinking it's not worth trying to eliminate the child paths from the set. For the case of individual files being added or deleted we won't get any benefit, because they're files and not directories so they don't have any children. And for directories being added or deleted I think it's rarer and doesn't save us that much compared to checking that the directory exists multiple times.

So I'm going to convert this to a regular `Set`. I think the `FilePathDiscovery` tests should then be good enough to show us it's still working.

---

Small aside. I wasn't sure what the behaviour is of a set when you're iterating over it using `for ... of` but then deleting elements and also maybe new elements get added while iterating. Turns out it handles that fine and it includes the new elements.
```
11:23:03 /Users/robertbrignull/github/vscode-codeql git:robertbrignull/resolve-queries-manual
> cat test.js
const xs = new Set();

xs.add(1);
xs.add(2);

for (const x of xs) {
	xs.delete(x);
	if (x === 1) {
		xs.add(3);
	}
	console.log(x);
}

console.log(xs);

11:23:19 /Users/robertbrignull/github/vscode-codeql git:robertbrignull/resolve-queries-manual
> node test.js
1
2
3
Set(0) {}
```